clang/include/clang/AST/Type.h

```diff
@@ -1126,9 +1126,6 @@ class QualType {
   /// Return true if this is a trivially relocatable type.
   bool isTriviallyRelocatableType(const ASTContext &Context) const;
 
-  /// Return true if this is a trivially equality comparable type.
-  bool isTriviallyEqualityComparableType(const ASTContext &Context) const;
```

Comment:
It's unfortunate that the other traits are implemented directly on `Type` but this one is in SemaExprCXX.cpp and isn't possible to compute without access to `Sema`. ISTM we should be able to answer this question on the type rather than requiring semantic analysis (considering potential uses like in clang-tidy where there is no `Sema` available).

Perhaps another way to approach this is with the `DefinitionData` for classes -- as we add members to the class, we build up information about whether something is trivially copyable, etc and store that on the bits defined in `CXXRecordDeclDefinitionBits.def`. Maybe we could do the same for equality comparable types, then we can leave this interface in `Type` and do special handling if the type is a `CXXRecordDecl`. WDYT?

Reply:
The main problem with this is that you don't have to have a member. Take this test case for example:
```c++
struct NotTriviallyEqualityComparableMoreConstrainedExternalOp {
  int i;
  bool operator==(const NotTriviallyEqualityComparableMoreConstrainedExternalOp&) const = default;
};

bool operator==(const NotTriviallyEqualityComparableMoreConstrainedExternalOp&,
                const NotTriviallyEqualityComparableMoreConstrainedExternalOp&) __attribute__((enable_if(true, ""))) {}

static_assert(!__is_trivially_equality_comparable(NotTriviallyEqualityComparableMoreConstrainedExternalOp));
```
I'm sure this can also be written with some `requires` clause. Without the builtin telling the library there is no way to figure out that a free function instead of the member is called.

Reply:
Well shoot! That's... frustrating. What about splitting the logic between `Type` and `Sema`? e.g., `Type` tracks whether the type *in isolation* is trivially equality comparable, and `Sema` can use that to early return if a type is not trivially equality comparable in isolation and do additional semantic checking otherwise to provide the value for the trait.

Reply:
We're not actually storing any infomation for `__is_trivially_equaltiy_comparable`, so I'm not sure what the benefit would be? If we ever do that we can of course split things up to say "this definitely isn't trivially equaility comparable".

Reply:
Yeah, it's not that we already have this information, it's a different approach to your current patch so that the functionality continues to live on `Type` (the benefit is that external consumers like downstreams or clang-tidy could continue to use the same interface they always have but get slightly more accurate results from it). If that is more effort than it's worth, I don't insist -- I'm not aware of any breakage from your current changes.

Reply:
Hmm. I'm not sure the is much use to "this is definitely not trivially equality comparable". Right now I think I'd rather just keep it in `SemaExprCXX`. If there comes up a use-case or we want to save some of the information we can still move some info into `Type`.

Reply:
Okay, I'm sold. Thanks for entertaining the idea!

Reply:
Always happy to do that!